### PR TITLE
Create Rust API Client

### LINKED ISSUE #7710: Add rust client to lakeFS release
As a first step - create auto generated lakeFS rust client
In a different issue - create a release process for the rust client with crate

### PR DESCRIPTION
**Pull request by N-o-Z**
Closes #7710

### Background

Create a lakeFS Rust client using Open API generator.
This PR only handles the code generation aspects.
Release workflow will be handled in a separate issue

### CODE REVIEW

**Comment by arielshaqed** [comment]
Please add the `.gitattributes` hack here, too.  Right now I am too scared to review.

**Comment by N-o-Z** [comment]
> Please add the `.gitattributes` hack here, too. Right now I am too scared to review.

Done

File: Makefile [hunk 1]
```diff
@@ -138,11 +140,20 @@ sdk-python: api/swagger.yml  ## Generate SDK for Python client - openapi generat
 		--package-name lakefs_sdk \
 		--http-user-agent "lakefs-python-sdk/$(PACKAGE_VERSION)" \
 		--git-user-id treeverse --git-repo-id lakeFS \
-		--additional-properties=infoName=Treeverse,infoEmail=services@treeverse.io,packageVersion=$(PACKAGE_VERSION),packageName=lakefs_sdk,packageVersion=$(PACKAGE_VERSION),projectName=lakefs-sdk,packageUrl=https://github.com/treeverse/lakeFS/tree/master/clients/python \
+		--additional-properties=infoName=Treeverse,infoEmail=services@treeverse.io,packageVersion=$(PACKAGE_VERSION),projectName=lakefs-sdk,packageUrl=https://github.com/treeverse/lakeFS/tree/master/clients/python \
```
Comment (arielshaqed) [inline Makefile:143]: Why this change?  We had two instances of "packageName", are you sure that the last change sticks?

**Review by arielshaqed — APPROVED** [review]
Thanks!

Just please add how we tested this generated client (hint: might look like "the same client worked when I generated it on my machine, and after I pull this code I will publish the crate and then test that that works")

File: Makefile [hunk 2]
```diff
@@ -12,6 +12,8 @@ OPENAPI_LEGACY_GENERATOR_IMAGE=openapitools/openapi-generator-cli:v5.3.0
 OPENAPI_LEGACY_GENERATOR=$(DOCKER) run --user $(UID_GID) --rm -v $(shell pwd):/mnt $(OPENAPI_LEGACY_GENERATOR_IMAGE)
 OPENAPI_GENERATOR_IMAGE=treeverse/openapi-generator-cli:v7.0.0.1
 OPENAPI_GENERATOR=$(DOCKER) run --user $(UID_GID) --rm -v $(shell pwd):/mnt $(OPENAPI_GENERATOR_IMAGE)
+OPENAPI_RUST_GENERATOR_IMAGE=openapitools/openapi-generator-cli:v7.5.0
```
[the remark this answers is not in the document]
Comment (Jonathan-Rosenberg) [reply Makefile:15]: Can we use the Treeverse image like in line 13?

File: clients/.gitattributes [hunk 3]
```diff
@@ -1,7 +1,8 @@
-java/**			linguist-generated
-java-legacy/**		linguist-generated
-python/**		linguist-generated
-python-legacy/**		linguist-generated
+java/**			    linguist-generated
+java-legacy/**	    linguist-generated
+python/**		    linguist-generated
```
Comment (Jonathan-Rosenberg) [inline clients/.gitattributes:3]: looks like a formatting issue

**Review by Jonathan-Rosenberg — APPROVED** [review]
Looks good although, as @arielshaqed mentioned, the Python changes should probably be included in another PR.
Added some non-blocking comments

File: clients/python-wrapper/tests/integration/test_object.py [hunk 4]
```diff
@@ -121,7 +121,7 @@ def test_object_copy(setup_repo):
     clt, repo = setup_repo
     data = "test_data"
     obj = WriteableObject(repository_id=repo.properties.id, reference_id="main", path="test_obj", client=clt).upload(
-        data=data, metadata={"foo": "bar"})
+        data=data, metadata={"foo": "bar", "key": "val"})
```
[the remark this answers is not in the document]
Comment (N-o-Z) [reply clients/python-wrapper/tests/integration/test_object.py:124]: Introduced by accident

File: Makefile (hunk 2 above)
[the remark this answers is not in the document]
Comment (N-o-Z) [reply Makefile:15]: I don't know the origin of 7.0.0.1. I'd rather use the latest and greatest for a new client.
I tried to modify the image for the java client but it made some minor semantic changes to the code which unfortunately we don't have a good way to test.
Therefore, I'd rather defer this and handle it when we do decide to remove all legacy clients

File: Makefile (hunk 1 above)
Comment (N-o-Z) [reply Makefile:143]: It was redundant, code generation didn't produce any changes

File: clients/.gitattributes (hunk 3 above)
Comment (N-o-Z) [reply clients/.gitattributes:3]: It's a visualization problem in github

**Comment by N-o-Z** [comment]
> Thanks!
> 
> Just please add how we tested this generated client (hint: might look like "the same client worked when I generated it on my machine, and after I pull this code I will publish the crate and then test that that works")

As you said - tested manually when build on my system.
We have a general issue with the generated clients and testing we should address at some point

File: Makefile (hunk 2 above)
[the remark this answers is not in the document]
Comment (arielshaqed) [reply Makefile:15]: AFAIK 7.0.0.1 _is_ the treeverse fork image.  AFAIU it corrects some issues in code generators for other languages.  Supporting multiple versions may be painful.

File: Makefile [hunk 5]
```diff
@@ -138,11 +140,20 @@ sdk-python: api/swagger.yml  ## Generate SDK for Python client - openapi generat
...
 
+sdk-rust: api/swagger.yml  ## Generate SDK for Rust client - openapi generator version 7.1.0
+	rm -rf clients/rust
+	mkdir -p clients/rust
+	$(OPENAPI_RUST_GENERATOR) generate \
+		-i /mnt/api/swagger.yml \
+		-g rust \
+		--additional-properties=infoName=Treeverse,infoEmail=services@treeverse.io,packageName=lakefs_sdk,packageVersion=$(PACKAGE_VERSION),packageUrl=https://github.com/treeverse/lakeFS/tree/master/clients/rust \
+		-o /mnt/clients/rust
+
```
Comment (arielshaqed) [inline Makefile:156]: This generated code is not rustfmt'ed.

We need either to:
* Add `--enable-post-process-file` here (and verify that it also works with the "rust" generator, because maybe it only works with "rust-server"), OR
* Add a "rustfmt" / "cargo fmt" command here to ensure that it's formatted (and then we will need to do something smarter about validation below!)

**Review by arielshaqed — COMMENTED** [review]
Another issue: unformatted Rust code is a pain.  Suggest fixing it here.

File: Makefile (hunk 5 above)
Comment (N-o-Z) [reply Makefile:156]: Not sure why we need to go to these lengths - it's auto generated code - if it compiles I don't care if it doesn't pass fmt
You can say the same for the python client which doesn't adhere to pep8 but we still don't perform any post processing

File: Makefile (hunk 2 above)
[the remark this answers is not in the document]
Comment (N-o-Z) [reply Makefile:15]: IMHO this version should be removed. 
1. We don't have the source for it so I don't know what were the local changes that were done
2. The issues this image was meant to correct might have been already fixed in the new version
3. We're not supporting multiple versions, we are using a single version to create a single client each time
4. If anything we need to remove the rogue image (i.e. treeverse/) and rely on an official image if we want to reduce the number of images

**Comment by ion-elgreco** [comment]
😲

File: Makefile (hunk 5 above)
Comment (N-o-Z) [reply Makefile:156]: After further checking, rustfmt takes a long time to run so I'm not sure it will be a good idea to add it to any of our makefile targets (build clients / checks-validator). Will merge as is for now and handle this later when this becomes an issue
